File: core/pkg/api/rack.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package api

import (
	"context"
	"go/types"

	"github.com/synnaxlabs/synnax/pkg/distribution/ontology"
	"github.com/synnaxlabs/synnax/pkg/service/access"
	"github.com/synnaxlabs/synnax/pkg/service/device"
	"github.com/synnaxlabs/synnax/pkg/service/rack"
	"github.com/synnaxlabs/synnax/pkg/service/status"
	"github.com/synnaxlabs/synnax/pkg/service/task"
	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/gorp"
	"github.com/synnaxlabs/x/validate"
)

type RackService struct {
	dbProvider
	accessProvider
	rack   *rack.Service
	device *device.Service
	task   *task.Service
	status statusProvider
}

func NewRackService(p Provider) *RackService {
	return &RackService{
		dbProvider:     p.db,
		rack:           p.Service.Rack,
		device:         p.Service.Device,
		task:           p.Service.Task,
		status:         p.status,
		accessProvider: p.access,
	}
}

type (
	Rack = rack.Rack
)

type (
	RackCreateRequest struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
	RackCreateResponse struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
)

func (svc *RackService) CreateRack(ctx context.Context, req RackCreateRequest) (res RackCreateResponse, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Create,
		Objects: rack.OntologyIDsFromRacks(req.Racks),
	}); err != nil {
		return res, err
	}
	if err := svc.WithTx(ctx, func(tx gorp.Tx) error {
		w := svc.rack.NewWriter(tx)
		for i, r := range req.Racks {
			if err := w.Create(ctx, &r); err != nil {
				return err
			}
			req.Racks[i] = r
		}
		res.Racks = req.Racks
		return nil
	}); err != nil {
		return RackCreateResponse{}, err
	}
	return res, nil
}

type (
	RackRetrieveRequest struct {
		Keys          []rack.Key `json:"keys" msgpack:"keys"`
		Names         []string   `json:"names" msgpack:"names"`
		SearchTerm    string     `json:"search_term" msgpack:"search_term"`
		Embedded      *bool      `json:"embedded" msgpack:"embedded"`
		HostIsNode    *bool      `json:"host_is_node" msgpack:"host_is_node"`
		Limit         int        `json:"limit" msgpack:"limit"`
		Offset        int        `json:"offset" msgpack:"offset"`
		IncludeStatus bool       `json:"include_status" msgpack:"include_status"`
	}
	RackRetrieveResponse struct {
		Racks []rack.Rack `json:"racks" msgpack:"racks"`
	}
)

func (svc *RackService) RetrieveRack(ctx context.Context, req RackRetrieveRequest) (res RackRetrieveResponse, _ error) {
	var (
		hasSearch = len(req.SearchTerm) > 0
		hasKeys   = len(req.Keys) > 0
		hasNames  = len(req.Names) > 0
		hasLimit  = req.Limit > 0
		hasOffset = req.Offset > 0
	)
	resRacks := make([]rack.Rack, 0, len(req.Keys)+len(req.Names))
	q := svc.rack.NewRetrieve()
	if hasKeys {
		q = q.WhereKeys(req.Keys...)
	}
	if hasNames {
		q = q.WhereNames(req.Names...)
	}
	if hasSearch {
		q = q.Search(req.SearchTerm)
	}
	if hasLimit {
		q = q.Limit(req.Limit)
	}
	if hasOffset {
		q = q.Offset(req.Offset)
	}
	if req.Embedded != nil {
		q = q.WhereEmbedded(*req.Embedded)
	}
	if req.HostIsNode != nil {
		q = q.WhereNodeIsHost(*req.HostIsNode)
	}
	if err := q.Entries(&resRacks).Exec(ctx, nil); err != nil {
		return res, err
	}

	if req.IncludeStatus {
		keys := make([]rack.Key, len(resRacks))
		for i := range resRacks {
			keys[i] = resRacks[i].Key
		}
		statuses := make([]rack.Status, 0, len(resRacks))
		if err := status.NewRetrieve[rack.StatusDetails](svc.status.status).
			WhereKeys(ontology.IDsToString(rack.OntologyIDsFromRacks(resRacks))...).
			Entries(&statuses).
			Exec(ctx, nil); err != nil {
			return res, err
		}
		for i, stat := range statuses {
			resRacks[i].Status = &stat
		}
	}

	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Retrieve,
		Objects: rack.OntologyIDsFromRacks(resRacks),
	}); err != nil {
		return res, err
	}
	res.Racks = resRacks
	return res, nil
}

type RackDeleteRequest struct {
	Keys []rack.Key `json:"keys" msgpack:"keys"`
}

func embeddedGuard(_ gorp.Context, r Rack) error {
	if !r.Embedded {
		return nil
	}
	return errors.Wrapf(validate.Error, "cannot delete embedded rack")
}

func (svc *RackService) DeleteRack(
	ctx context.Context,
	req RackDeleteRequest,
) (res types.Nil, _ error) {
	if err := svc.access.Enforce(ctx, access.Request{
		Subject: getSubject(ctx),
		Action:  access.Delete,
		Objects: rack.OntologyIDs(req.Keys),
	}); err != nil {
		return res, err
	}
	return res, svc.WithTx(ctx, func(tx gorp.Tx) error {
		exists, err := svc.device.NewRetrieve().WhereRacks(req.Keys...).Exists(ctx, tx)
		if err != nil {
			return err
		}
		if exists {
			return errors.Wrapf(validate.Error, "cannot delete rack when devices are still attached")
		}
		exists, err = svc.task.NewRetrieve().WhereInternal(false, gorp.Required()).WhereRacks(req.Keys...).Exists(ctx, tx)
		if err != nil {
			return err
		}
		if exists {
			return errors.Wrapf(validate.Error, "cannot delete rack when tasks are still attached")
		}
		w := svc.rack.NewWriter(tx)
		for _, k := range req.Keys {
			if err = w.DeleteGuard(ctx, k, embeddedGuard); err != nil {
				return err
			}
		}
		return nil
	})
}

File: client/cpp/rack/rack.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "client/cpp/errors/errors.h"
#include "client/cpp/rack/rack.h"
#include "client/cpp/task/task.h"
#include "x/cpp/errors/errors.h"

namespace synnax::rack {
std::pair<Rack, x::errors::Error> Rack::from_proto(const api::v1::Rack &rack) {
    Rack r;
    r.key = rack.key();
    r.name = rack.name();
    if (rack.has_status()) {
        auto [s, err] = Status::from_proto(rack.status());
        if (err) return {r, err};
        r.status = s;
    }
    return {r, x::errors::NIL};
}

void Rack::to_proto(api::v1::Rack *rack) const {
    rack->set_key(key);
    rack->set_name(name);
    if (!status.is_zero()) status.to_proto(rack->mutable_status());
}

Client::Client(
    std::unique_ptr<CreateClient> rack_create_client,
    std::unique_ptr<RetrieveClient> rack_retrieve_client,
    std::unique_ptr<DeleteClient> rack_delete_client,
    std::shared_ptr<task::CreateClient> task_create_client,
    std::shared_ptr<task::RetrieveClient> task_retrieve_client,
    std::shared_ptr<task::DeleteClient> task_delete_client
):
    rack_create_client(std::move(rack_create_client)),
    rack_retrieve_client(std::move(rack_retrieve_client)),
    rack_delete_client(std::move(rack_delete_client)),
    task_create_client(std::move(task_create_client)),
    task_retrieve_client(std::move(task_retrieve_client)),
    task_delete_client(std::move(task_delete_client)) {}

std::pair<Rack, x::errors::Error> Client::retrieve(const rack::Key key) const {
    auto req = api::v1::RackRetrieveRequest();
    req.add_keys(key);
    auto [res, err] = rack_retrieve_client->send("/rack/retrieve", req);
    if (err) return {Rack(), err};
    if (res.racks_size() == 0)
        return {Rack(), errors::not_found_error("Rack", "key " + std::to_string(key))};
    auto [rack, proto_err] = Rack::from_proto(res.racks(0));
    if (proto_err) return {Rack(), proto_err};
    rack.tasks = task::Client(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return {rack, x::errors::NIL};
}

std::pair<Rack, x::errors::Error> Client::retrieve(const std::string &name) const {
    auto req = api::v1::RackRetrieveRequest();
    req.add_names(name);
    auto [res, err] = rack_retrieve_client->send("/rack/retrieve", req);
    if (err) return {Rack(), err};
    if (res.racks_size() == 0)
        return {Rack(), errors::not_found_error("Rack", "name " + name)};
    if (res.racks_size() > 1)
        return {Rack(), errors::multiple_found_error("racks", "name " + name)};
    auto [rack, proto_err] = Rack::from_proto(res.racks(0));
    if (proto_err) return {Rack(), proto_err};
    rack.tasks = task::Client(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return {rack, x::errors::NIL};
}

x::errors::Error Client::create(Rack &rack) const {
    auto req = api::v1::RackCreateRequest();
    rack.to_proto(req.add_racks());
    auto [res, err] = rack_create_client->send("/rack/create", req);
    if (err) return err;
    if (res.racks_size() == 0) return errors::unexpected_missing_error("rack");
    rack.key = res.racks().at(0).key();
    rack.tasks = task::Client(
        rack.key,
        task_create_client,
        task_retrieve_client,
        task_delete_client
    );
    return err;
}

std::pair<Rack, x::errors::Error> Client::create(const std::string &name) const {
    auto rack = Rack{.name = name};
    auto err = create(rack);
    return {rack, err};
}

x::errors::Error Client::del(const rack::Key key) const {
    auto req = api::v1::RackDeleteRequest();
    req.add_keys(key);
    auto [res, err] = rack_delete_client->send("/rack/delete", req);
    return err;
}
}
